circuits-lib/src/bridge_circuit/groth16.rs

```diff
@@ -1,94 +1,154 @@
-
-pub fn create_output_digest(total_work: &[u8; 16]) -> [u8; 32] {
-    let total_work_digest: [u8; 32] = Sha256::digest(total_work).into();
-    let len_output: [u8; 2] = hex::decode("0200").unwrap().try_into().unwrap();
-
-    let output_pre_digest: [u8; 98] = [
-        &OUTPUT_TAG,
-        &total_work_digest[..],
-        &ASSUMPTIONS[..],
-        &len_output[..],
-    ]
-    .concat()
-    .try_into()
-    .expect("slice has correct length");
-
-    Sha256::digest(output_pre_digest).into()
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize, Compress, SerializationError};
+type G1 = ark_bn254::G1Affine;
+type G2 = ark_bn254::G2Affine;
+
+#[derive(Copy, Clone, Debug)]
+pub struct CircuitGroth16Proof {
+    a: G1,
+    b: G2,
+    c: G1,
 }
 
-pub fn create_claim_digest(output_digest: &[u8; 32], pre_state: &[u8; 32]) -> [u8; 32] {
-    let data: [u8; 8] = [0; 8];
+impl CircuitGroth16Proof {
+    pub fn new(a: G1, b: G2, c: G1) -> CircuitGroth16Proof {
+        CircuitGroth16Proof { a, b, c }
+    }
 
-    let claim_len: [u8; 2] = [4, 0];
+    pub fn from_seal(seal: &[u8; 256]) -> CircuitGroth16Proof {
+        let a = G1::new(
+            ark_bn254::Fq::from_be_bytes_mod_order(&seal[0..32]),
+            ark_bn254::Fq::from_be_bytes_mod_order(&seal[32..64]),
+        );
+
+        let b = G2::new(
+            ark_bn254::Fq2::from_base_prime_field_elems([
+                ark_bn254::Fq::from_be_bytes_mod_order(&seal[96..128]),
+                ark_bn254::Fq::from_be_bytes_mod_order(&seal[64..96]),
+            ])
+            .unwrap(),
+            ark_bn254::Fq2::from_base_prime_field_elems([
+                ark_bn254::Fq::from_be_bytes_mod_order(&seal[160..192]),
+                ark_bn254::Fq::from_be_bytes_mod_order(&seal[128..160]),
+            ])
+            .unwrap(),
+        );
+
+        let c = G1::new(
+            ark_bn254::Fq::from_be_bytes_mod_order(&seal[192..224]),
+            ark_bn254::Fq::from_be_bytes_mod_order(&seal[224..256]),
+        );
+
+        CircuitGroth16Proof::new(a, b, c)
+    }
 
-    let concatenated = [
-        &CLAIM_TAG,
-        &INPUT,
-        pre_state,
-        &POST_STATE,
-        output_digest,
-        &data[..],
-        &claim_len,
-    ]
-    .concat();
+    pub fn from_compressed(
+        compressed: &[u8; 128],
+    ) -> Result<CircuitGroth16Proof, SerializationError> {
+        let a_compressed = &compressed[0..32];
+        let b_compressed = &compressed[32..96];
+        let c_compressed = &compressed[96..128];
+        let a = ark_bn254::G1Affine::deserialize_compressed(a_compressed)?;
+        let b = ark_bn254::G2Affine::deserialize_compressed(b_compressed)?;
+        let c = ark_bn254::G1Affine::deserialize_compressed(c_compressed)?;
+
+        Ok(CircuitGroth16Proof::new(a, b, c))
+    }
 
-    let mut claim_digest = Sha256::digest(concatenated);
-    claim_digest.reverse();
+    pub fn to_compressed(&self) -> Result<[u8; 128], SerializationError> {
```

Comment:
Can this actually fail? If we're validating all elements when creating, why do we need to return a result here? If this cannot fail, let's replace unwraps with `except` and explain why it can't fail. If this can fail, then let's return the error instead of unwrapping